base/channels.jl

```diff
@@ -454,6 +455,156 @@ function put_unbuffered(c::Channel, v)
     return v
 end
 
+"""
+    append!(c::Channel, iter)
+
+Append all items in `iter` to the channel `c`. If the channel is buffered, this operation requires
+fewer `lock` operations than individual `put!`s. Blocks if the channel is full.
+
+# Examples
+
+```jldoctest
+julia> c = Channel(3);
+
+julia> append!(c, 1:3);
+
+julia> take!(c)
+1
+
+julia> take!(c)
+2
+
+julia> take!(c)
+3
+```
+
+!!! compat "Julia 1.12"
+    Requires at least Julia 1.12.
+"""
+function append!(c::Channel, iter::T) where {T}
+    if hasmethod(length, Tuple{T})
+        len = length(iter)
+        # shortcircuit for small short objects
+        if len == 0
+            return c
+        elseif len == 1
+            put!(c, @inbounds vec[begin])
+            return c
+        end
+    end
+
+    return isbuffered(c) ? append_buffered(c, iter) : append_unbuffered(c, iter)
+end
+function append!(c1::Channel, c2::Channel{T}) where {T}
+    # if the source channel has no buffer or only one slot, there is no benefit to grouping the lock
+    if min(c1.sz_max, c2.sz_max) < 2
+        return isbuffered(c1) ? append_buffered(c1, c2) : append_unbuffered(c1, c2)
+    end
+    i = 0
+    lock(c2)
+    try
+        while isopen(c2) || isready(c2)
+            # wait for data to become available
+            while isempty(c2.data)
+                try
+                    check_channel_state(c2)
+                    wait(c2.cond_take)
+                catch e
+                    if isa(e, InvalidStateException) && e.state === :closed
+                        break
+                    else
+                        rethrow()
+                    end
+                end
+            end
+
+            lock(c1)
+            try
+                while length(c1.data) == c1.sz_max
+                    check_channel_state(c1)
+                    wait(c1.cond_put)
+                end
+
+                # append as many as possible to c1
+                n = min(c1.sz_max - length(c1.data), length(c2.data))
+                _increment_n_avail(c1, n)
+                c1_len_og = length(c1.data)
+
+                resize!(c1.data, c1_len_og + n)
+                copyto!(c1.data, c1_len_og + 1, c2.data, 1, n)
+                deleteat!(c2.data, 1:n)
+
+                _increment_n_avail(c2, -n)
+                notify(c1.cond_take, nothing, true, false)
+                foreach(_ -> notify(c2.cond_put, nothing, true, false), 1:n)
+            finally
+                unlock(c1)
+            end
+        end
+    finally
+        unlock(c2)
+    end
+    return c1
+end
+
+function append_unbuffered(c::Channel, iter)
+    for v in iter
+        put!(c, v)
+    end
+    return c
+end
+
+function append_buffered(c::Channel{T}, iter::I) where {T,I}
+    converted_items = Iterators.Stateful(Iterators.map(x -> convert(T, x), iter))
+
+    has_length = hasmethod(length, Tuple{I})
```

Comment:
Need to replace this check with `Base.IteratorSize` once I get feedback from @jakobnissen